plugin-niassa/src/main/java/ca/on/oicr/gsi/shesmu/niassa/MigrationAction.java

```diff
@@ -0,0 +1,385 @@
+            String provider;
+            if (limsKey.getProvider().equals("pinery-miso")
+                || limsKey.getProvider().equals("pinery-miso-2.2")) {
+              provider = "pinery-miso";
+            } else {
+              int index = limsKey.getProvider().indexOf("-v");
+              provider = limsKey.getProvider().substring(0, index);
+            }
+
+            ExternalKey key =
+                new ExternalKey(
+                    provider, limsKey.getId(), Map.of("pinery-hash-7", limsKey.getVersion()));
+            externalKeys.add(key);
+          });
+
+      // Populate arguments
+      final var vidarrDBUrl = configuration.map(Configuration::getVidarrDBUrl).orElse(null);
+      final var user = configuration.map(Configuration::getVidarrDBUser).orElse(null);
+      final var password = configuration.map(Configuration::getVidarrDBPassword).orElse(null);
+
+      // Get hash id's for all input file SWIDs from matches and populate metadata
+      if (vidarrDBUrl != null) {
+        try (Connection conn = DriverManager.getConnection(vidarrDBUrl, user, password)) {
+          Statement statement = conn.createStatement();
+          final var inputFiles = matches.get(0).state().inputFiles();
+          final var workflowRunSWID = matches.get(0).state().workflowRunAccession();
+          final var workflowRunEssentials = server.get().directoryAndIni(workflowRunSWID);
+
+          ResultSet rs =
+              statement.executeQuery(
+                  "SELECT hash_id FROM analysis WHERE labels->'file-accession' "
+                      + "IN ("
+                      + inputFiles.stream().map(String::valueOf).collect(Collectors.joining(","))
+                      + ")");
+
+          // No rows means input workflow hasn't been migrated yet so wait
+          if (!rs.next()) {
+            return ActionState.WAITING;
+          }
+
+          final var arrayNode = MAPPER.createArrayNode();
+          while (rs.next()) {
+            String hashId = rs.getString("hash_id");
+            String contents = "{\"type\": \"INTERNAL\", \"contents\":\"vidarr:_/" + hashId + "\"}";
+            JsonNode node = MAPPER.readTree(contents);
+            arrayNode.add(node);
+          }
+
+          arguments.put("workflowRunSWID", workflowRunSWID);
+          arguments.putPOJO("inputFiles", arrayNode);
+          arguments.put("ini", String.valueOf(workflowRunEssentials.ini()));
+        } catch (SQLException | JsonProcessingException e) {
+          e.printStackTrace();
+          return ActionState.FAILED; // Should this fail?
+        }
+      }
+
+      // Populate metadata
+      final var outputFileMigration = MAPPER.createArrayNode();
+      final var outputFiles = matches.get(0).state().files();
+      outputFiles.forEach(
+          file -> {
+            var provider = file.iterator().next().first();
+            var id = file.iterator().next().second();
+            String tree =
+                "{\"fileSWID\": "
```

Comment:
I would use the `MAPPER` to create this programmatically. There can be multiple LIMS keys per file, so you'll have to construct an array of them.